test/forms/compoundField.test.js

```diff
@@ -0,0 +1,195 @@
+    it("serializes it's child fields in to an object", () => {
+      const date = new CompoundField('date',
+        textField('day'),
+        textField('month'),
+        textField('year')
+      );
+      date.day.value = '6';
+      date.month.value = '12';
+      date.year.value = '2017';
+
+      const obj = date.serialize();
+      expect(obj).to.eql({ date: { day: '6', month: '12', year: '2017' } });
+    });
+  });
+
+  describe('#validate', () => {
+    it('returns true if validations pass', () => {
+      const date = new CompoundField('date',
+        textField('day').joi('day error', Joi.number()),
+        textField('month').joi('month error', Joi.number()),
+        textField('year').joi('year error', Joi.number())
+      ).joi('date error', Joi.object().and('day', 'month', 'year'));
+
+      date.parse({ day: 1, month: 12, year: 1988 });
+      expect(date.validate()).to.be.true;
+    });
+
+    it("returns false if it's validations fail", () => {
+      const parentWillFail = new CompoundField(
+        'parent',
+        textField('child')
+      ).joi('will fail', Joi.object().keys({ child: Joi.required() }));
+
+      expect(parentWillFail.validate()).to.be.false;
+    });
+
+    it('returns false if a child field validation fails', () => {
+      const childWillFail = new CompoundField(
+        'parent',
+        textField('child').joi('will fail', Joi.required())
+      );
+
+      expect(childWillFail.validate()).to.be.false;
+    });
+
+    it('executes any validations on the child fields', () => {
+      const error = name => `${name} is required`;
+
+      const date = new CompoundField('date',
+        textField('day').joi(error('day'), Joi.number().required()),
+        textField('month').joi(error('month'), Joi.number().required()),
+        textField('year').joi(error('year'), Joi.number().required())
+      );
+      date.validate();
+      expect(date.day.errors).to.contain(error('day'));
+      expect(date.month.errors).to.contain(error('month'));
+      expect(date.year.errors).to.contain(error('year'));
+    });
+
+    it("executes it's validations", () => {
+      const date = new CompoundField('date',
+        textField('day'),
+        textField('month'),
+        textField('year')
+      ).joi(
+        'A date is required',
+        Joi.object().and('day', 'month', 'year')
+      );
+
+      date.parse({ day: 10 });
+      date.validate();
+      expect(date.errors).to.contain('A date is required');
+    });
+
+    it('attaches a targeted error to the correct child field', () => {
+      const date = new CompoundField('date',
+        textField('day'),
+        textField('month'),
+        textField('year')
+      ).joi(
+        errorFor('day', 'Day is required'),
+        Joi.object()
+          .with('year', 'day')
+          .with('month', 'day')
+      );
+
+      date.parse({ year: 2017 });
+      date.validate();
+      expect(date.day.errors).to.contain('Day is required');
```

Comment:
only testing the one?

Reply:
I've definitely under tested compoundField in favour of getting it done and out to SSCS in time for their deadline.

TBH I'm not happy with the value mapping and I don't like how I'm copying the validation logic around. This needs a refactor. Compound field feels very similar to a form (takes fields, validates fields, parses body by delegating to fields) so there's a lot of cleaning that could happen here.